File: morning_exercises/day20.cpp
#include <iostream>
#include <memory>
#include <deque>
#include <ostream>
// #include <queue>
#include <string>

class BinaryNode {
    public:
        int val = 0;
        std::shared_ptr<BinaryNode> left = nullptr;
        std::shared_ptr<BinaryNode> right = nullptr;
    
    BinaryNode(int val): val(val) {}

    static void serialiseHelper(std::shared_ptr<BinaryNode> root, std::deque<std::string>& res) {
        if (root==nullptr) {
            res.push_back("N");
            return;
        }

        res.push_back(std::to_string(root->val));
        serialiseHelper(root->left, res);
        serialiseHelper(root->right, res);
    }

    static std::deque<std::string> serialise(std::shared_ptr<BinaryNode> root) {
        std::deque<std::string> res = {};

        serialiseHelper(root, res);

        return res;
    }

    static std::shared_ptr<BinaryNode> deserialiseHelper(std::deque<std::string>& input) {
        std::string val = input.front();
        input.pop_front();

        if (val=="N") return nullptr;

        auto node = std::make_shared<BinaryNode>(std::stoi(val));
        node->left = deserialiseHelper(input);
        node->right = deserialiseHelper(input);
        return node;
    }

    static std::shared_ptr<BinaryNode> deserialise(std::deque<std::string> input) {
        return deserialiseHelper(input);
    }
};

int main() {
    auto root = std::make_shared<BinaryNode>(11);
    root->left = std::make_shared<BinaryNode>(2);
    root->right = std::make_shared<BinaryNode>(3);
    root->left->left = std::make_shared<BinaryNode>(4);
    root->left->right = std::make_shared<BinaryNode>(5);

    std::deque<std::string> res = BinaryNode::serialise(BinaryNode::deserialise(BinaryNode::serialise(root)));

    for (std::string& s : res) {
        std::cout << s << " ";
    }
    std::cout << std::endl;
}

File: morning_exercises/day20.cs
class BinaryNode5(int val = 0, BinaryNode5? left = null, BinaryNode5? right = null)
{
    public int val = val;
    public BinaryNode5? left = left;
    public BinaryNode5? right = right;

    public static void SerialiseHelperQA(BinaryNode5? root, Queue<string> vals)
    {
        if (root==null)
        {
            vals.Enqueue("N");
            return;
        }

        vals.Enqueue(root.val.ToString());

        // vals.Enqueue(root==null ? "N" : root.val.ToString());

        SerialiseHelperQA(root.left, vals);
        SerialiseHelperQA(root.right, vals);
    }

    public static Queue<string> SerialiseQA(BinaryNode5? root)
    {
        Queue<string> vals = new();

        SerialiseHelperQA(root, vals);
        return vals;
    }

    public static void SerialiseHelperLLA(BinaryNode5? root, LinkedList<string> input)
    {
        if (root==null) 
        {
            input.AddLast("N");
            return;
        }

        input.AddLast(root.val.ToString());

        SerialiseHelperLLA(root.left, input);
        SerialiseHelperLLA(root.right, input);
    }

    public static LinkedList<string> SerialiseLLA(BinaryNode5? root)
    {
        LinkedList<string> input = new();

        SerialiseHelperLLA(root, input);

        return input;
    }

    public static BinaryNode5? DeserialiseQA(Queue<string> res)
    {
        string token = res.Dequeue();
        if (token == "N") return null;
        // int? val = token=="N" ? null : int.Parse(token);

        BinaryNode5? node = new(int.Parse(token))
        {
            left = DeserialiseQA(res),
            right = DeserialiseQA(res)
        };
        return node;
    }

    public static BinaryNode5? DeserialiseLLA(LinkedList<string> input)
    {
        string val = input.First();
        // string val = input.First!.Value;
        input.RemoveFirst();

        if (val=="N") return null;

        BinaryNode5? node = new(int.Parse(val))
        {
            left = DeserialiseLLA(input),
            right = DeserialiseLLA(input)
        };
        return node;
    }
}